commit f062a7ad50b4308413007148795cc784c158d7bd
Author: John Leach <john@johnleach.co.uk>
Date:   Fri Mar 15 11:50:44 2019 +0000

    This fixes several tests that are broken when the SOURCE_DATE_EPOCH environment
    variable is set. TestGemPackageTarWriter was a bit complicated so I just unset
    the env var for the duration of the tests, and restored it.

diff --git a/test/rubygems/test_gem_specification.rb b/test/rubygems/test_gem_specification.rb
index d2ecbf44..915afab9 100644
--- a/test/rubygems/test_gem_specification.rb
+++ b/test/rubygems/test_gem_specification.rb
@@ -95,6 +95,8 @@ def setup
 
     @current_version = Gem::Specification::CURRENT_SPECIFICATION_VERSION
 
+    @gem_spec_today = ENV["SOURCE_DATE_EPOCH"] ? Time.at(ENV["SOURCE_DATE_EPOCH"].to_i) : Gem::Specification::TODAY
+
     load 'rubygems/syck_hack.rb'
   end
 
@@ -1033,7 +1035,8 @@ def test_self_load_legacy_ruby
     end
     assert_equal 'keyedlist', spec.name
     assert_equal '0.4.0', spec.version.to_s
-    assert_equal Gem::Specification::TODAY, spec.date
+    spec_date = @gem_spec_today
+    assert_equal Time.utc(spec_date.year, spec_date.month, spec_date.day), spec.date
     assert spec.required_ruby_version.satisfied_by?(Gem::Version.new('1'))
     assert_equal false, spec.has_unit_tests?
   end
@@ -1674,7 +1677,10 @@ def test_contains_requirable_file_eh_extension
   end
 
   def test_date
+    backup_epoch = ENV["SOURCE_DATE_EPOCH"]
+    ENV["SOURCE_DATE_EPOCH"] = nil
     assert_equal Gem::Specification::TODAY, @a1.date
+    ENV["SOURCE_DATE_EPOCH"] = backup_epoch
   end
 
   def test_date_equals_date
@@ -2390,7 +2396,7 @@ def test_to_ruby
   s.required_rubygems_version = Gem::Requirement.new(\"> 0\".freeze) if s.respond_to? :required_rubygems_version=
   s.require_paths = ["lib".freeze, "other".freeze]
   s.authors = ["A User".freeze]
-  s.date = "#{Gem::Specification::TODAY.strftime "%Y-%m-%d"}"
+  s.date = "#{@gem_spec_today.strftime "%Y-%m-%d"}"
   s.description = "This is a test description".freeze
   s.email = "example@example.com".freeze
   s.files = ["lib/code.rb".freeze]
@@ -2439,7 +2445,7 @@ def test_to_ruby_for_cache
   s.required_rubygems_version = Gem::Requirement.new(\"> 0\".freeze) if s.respond_to? :required_rubygems_version=
   s.require_paths = ["lib".freeze]
   s.authors = ["A User".freeze]
-  s.date = "#{Gem::Specification::TODAY.strftime "%Y-%m-%d"}"
+  s.date = "#{@gem_spec_today.strftime "%Y-%m-%d"}"
   s.description = "This is a test description".freeze
   s.email = "example@example.com".freeze
   s.homepage = "http://example.com".freeze
@@ -2496,7 +2502,7 @@ def test_to_ruby_fancy
   s.required_rubygems_version = Gem::Requirement.new(\">= 0\".freeze) if s.respond_to? :required_rubygems_version=
   s.require_paths = ["lib".freeze]
   s.authors = ["A User".freeze]
-  s.date = "#{Gem::Specification::TODAY.strftime "%Y-%m-%d"}"
+  s.date = "#{@gem_spec_today.strftime "%Y-%m-%d"}"
   s.description = "This is a test description".freeze
   s.email = "example@example.com".freeze
   s.executables = ["exec".freeze]
@@ -3576,7 +3582,7 @@ def test_metadata_specs
   s.metadata = { "one" => "two", "two" => "three" } if s.respond_to? :metadata=
   s.require_paths = ["lib".freeze]
   s.authors = ["A User".freeze]
-  s.date = "#{Gem::Specification::TODAY.strftime("%Y-%m-%d")}"
+  s.date = "#{@gem_spec_today.strftime("%Y-%m-%d")}"
   s.description = "This is a test description".freeze
   s.email = "example@example.com".freeze
   s.files = ["lib/code.rb".freeze]
